logic/pulsed/sequence_generator_logic.py

```diff
@@ -1896,6 +1913,11 @@ def sample_pulse_sequence(self, sequence):
 
                 # Add created waveform names to the set
                 written_waveforms.update(waveform_list)
+            else:
+                self.log.debug('Waveform already sampled: {0}'.format(name_tag))
+                ensemble_info = self.get_ensemble(name_tag).sampling_information.copy()
```

Comment:
The `ensemble_info` dict returned by `sample_pulse_block_ensemble` is not the same as `sampling_information` inside the ensemble instances. They share some information and it could even be that one is a subset of the other but they must be consistent either way. So do not simply copy but ensure that the dict contains exactly the same (and only the) information as would be returned by `sample_pulse_block_ensemble`.

Reply:
I really like the idea, but is it possible to make it optional? It often happens to me that I run the same experiment several times with different parameters. So I want some of the ensembles which have been sampled previously to be overwritten. If I understand correctly I would then need to rename the new individual ensembles or I have to clear the old ones by hand or I have to generate the new ensembles in a seperate step...
Of course it would be optimal if I can decide before which ensembles to keep (and only sample if not present) and which to sample in any case, but I am not sure how to realise that

Reply:
@SimonSchmitt88:
Depending on HOW you introduce changes to an existing `PulseBlockEnsemble` this method would still sample it anew. 
Let's assume you have `myens` which is an instance of `PulseBlockEnsemble` and you want to set a new `PulseBlock` named `'myblk'` with 0 reps at index 1 by calling the following:
```python
myens[1] = ('myblk', 0)
```
If `myens` has been sampled before that, this setting of a new `PulseBlock` at index 1 will automatically trigger new sampling of the waveform. Actually it will every time you alter its items directly.

This is because you can handle these pulse objects like lists and every time the method `__setitem__` gets called (e.g. by the example above) it will delete the `sampling_information` dict inside the `PulseBlockEnsemble`. And this in turn will cause it to be assumed not sampled.
If you alter the `PulseBlock` instances directly, this will not be noticed.

I think the easiest solution for your scripts would be one of:

- Manually set `myens.sampling_information = dict()` if you alter `myens`
- Call the example above if you have altered `myblk`
- Simply regenerate the affected ensemble if you want to alter it (this is the intended method). The pure generation of a pulsed object is negligibly fast. I would strongly recommend this way because it is also the one to be guaranteed working in future qudi versions.

Reply:
deleted the `pulse_generator_settings` from the `ensemble_info`, as they are not in the `ensemble_info` for freshly sampled ensembles. The `written_waveforms` are left in, as they are added afterwards by hand anyways.